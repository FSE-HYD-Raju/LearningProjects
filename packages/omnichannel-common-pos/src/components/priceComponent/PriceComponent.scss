@import '../../variables';

.pr {
	@media (min-width: 1033px) {
		&-container {
			border-left: 1px solid #eee;
		}
	}

	.StandardPriceRenderer.pr-container,
	&-container {
		display: flex;
		flex: 1 1 auto;
		flex-direction: column;
		padding: 1.875rem 0 1.25rem 0.9375rem;
		color: darkgray !important;
		justify-content: center;
	}

	&-text-block {
		display: flex;
		align-self: center;
	}

	&-price-line {
		display: flex;
		flex: 1 1 auto;
		align-items: stretch;
		justify-content: center;
		width: 100%;
	}

	&-price-title {
		align-self: flex-start;
		height: 4rem;
		font-weight: bold;
		align-items: center;
		display: flex;
		flex: 0 0 6rem;
	}

	&-price-value {
		color: $green;
		font-size: $h4-font-size;
		font-weight: 700;
		align-self: flex-end;
		flex: 1 1 auto;
		height: 4rem;
		display: flex;
		align-items: center;
		justify-content: flex-end;
		padding-right: 6px;
	}

	&-price-value.upfront {
		font-size: $h3-font-size;
	}

	&-integer {
		font-size: $h1-font-size;
		font-weight: 600;
	}

	&-currency {
		color: $green;
	}

	&-fraction-currency {
		font-weight: 600;
	}

	&-payment-type-text {
		font-size: $font-size-sm;
		font-weight: $font-weight-normal;
		text-transform: uppercase;
	}

	&-fraction {
		display: flex;
		flex-direction: column;
		justify-content: center;
		padding-left: 0.1875rem;
	}

	&-prices-block {
		display: flex;
		flex: initial;
		flex-wrap: nowrap;
		flex-direction: column;
	}
}
