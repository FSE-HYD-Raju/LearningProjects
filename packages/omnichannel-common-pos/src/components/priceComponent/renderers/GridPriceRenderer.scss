@import '../../../variables';

.GridPriceRenderer {
	display: flex;
	align-items: center;
	justify-content: center;
	flex-wrap: wrap;

	&-price-line {
		display: flex;
		justify-content: center;
		align-items: center;
		margin: 0 .5rem;
	}

	&-price-value {
		font-size: $font-size-lg;
		font-weight: $headings-font-weight;
		display: flex;
		align-items: center;
	}

	&-payment-type-text {
		color: $gray-900;
		padding-left: 5px;
	}
}
