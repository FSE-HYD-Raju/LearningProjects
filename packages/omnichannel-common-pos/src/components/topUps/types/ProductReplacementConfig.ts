interface ProductReplacementConfig {
	individualId?: string;
	agreementId?: string;
	productId?: string;
	productOfferingId?: string;
	replaceType?: string;
	configuration?: object;
}
export { ProductReplacementConfig };
