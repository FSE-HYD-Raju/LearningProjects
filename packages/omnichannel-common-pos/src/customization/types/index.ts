import { CustomizationBootstrapType, GetStoreConfig } from "./CustomizationBootstrapType";
import { CustomizationClass } from "./CustomizationClass";
import { CustomizationFunction } from "./CustomizationFunction";
import { CustomizationPointsMapping } from "./CustomizationPointsMapping";

export {
	GetStoreConfig,
	CustomizationBootstrapType,
	CustomizationClass,
	CustomizationFunction,
	CustomizationPointsMapping
};
