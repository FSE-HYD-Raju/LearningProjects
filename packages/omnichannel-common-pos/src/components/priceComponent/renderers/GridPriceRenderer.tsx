import * as React from "react";
import { PriceRendererProps } from "../PriceComponent";
import FormattedMessage from "../../../channelUtils/FormattedMessage";
import FormattedNumber from "../../../channelUtils/FormattedNumber";
import { PriceComponentPriceType } from "../model/PriceComponentPriceType";

const GridPriceRenderer: React.FC<PriceRendererProps> = (props) => {
	return (
		<div className="GridPriceRenderer">
			{props.prices.map((price: PriceComponentPriceType) => {
				if (!price.integer || !price.fraction || !price.currency) {
					return null;
				}
				const isUpfront = price.priceType === "upfront";
				return (
					<div className="GridPriceRenderer-price-line" key={price.priceType}>
						<div className="GridPriceRenderer-price-value">
							<FormattedNumber
								style="currency"
								currency={price.currency}
								value={props.priceValueCalculator(price)}
								minimumFractionDigits={2}
								maximumFractionDigits={2}
							/>
						</div>
						<div className="GridPriceRenderer-payment-type-text">
							{!isUpfront && <FormattedMessage {...props.messages.monthly}/>}
							{isUpfront && <FormattedMessage {...props.messages.now}/>}
						</div>
					</div>
				);
			})}
		</div>
	);
};

export default GridPriceRenderer;
