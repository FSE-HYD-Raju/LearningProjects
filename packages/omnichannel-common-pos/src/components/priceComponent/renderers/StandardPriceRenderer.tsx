import * as React from "react";
import { PriceRendererProps } from "../PriceComponent";
import FormattedMessage from "../../../channelUtils/FormattedMessage";
import FormattedNumber from "../../../channelUtils/FormattedNumber";
import { PriceComponentPriceType } from "../model/PriceComponentPriceType";

const StandardPriceRenderer: React.FC<PriceRendererProps> = (props) => {
	return (
		<div className="StandardPriceRenderer pr-container">
			<div className="pr-text-block">
				<FormattedMessage {...props.messages.starting}/>
			</div>
			<div className="pr-prices-block">
				{props.prices.map((price: PriceComponentPriceType) => {
					if (!price.integer || !price.fraction || !price.currency) {
						return null;
					}
					return (
						<div className="pr-price-line" key={price.priceType}>
							<div className="pr-integer pr-currency">
								<FormattedNumber
									style="currency"
									currency={price.currency}
									value={parseInt(price.integer, 10)} // due to fraction displayed separately
									minimumFractionDigits={0}
									maximumFractionDigits={0}
								/>
							</div>
							<div className="pr-fraction">
								<div className="pr-currency pr-fraction-currency">
									{price.fraction}
								</div>
								<div className="pr-payment-type-text">
									{price.priceType === "upfront"
										? (<FormattedMessage {...props.messages.now}/>)
										: (<FormattedMessage {...props.messages.mnth}/>)}
								</div>
							</div>
						</div>
					);
				})}
			</div>
		</div>
	);
};

export default StandardPriceRenderer;
