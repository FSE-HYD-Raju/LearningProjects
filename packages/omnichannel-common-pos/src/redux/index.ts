"use strict";

import { default as actions, AppActions } from "./actions";
import { default as reducers, AppState } from "./reducers";
import { InterceptReducers } from "./reducers/reducerUtils";

import { default as DevTools } from "./DevTools";

import { connectFluxRedux, getDefaultFluxStoreMap } from "./connectFluxRedux";
import { configureStore, AppStore, StoreConfig } from "./store";
import * as actionTypes from "./actions";
import * as decorators from "./decorators";

import * as channelUtils from "./utils/channel";
export * from "./utils";
import * as Selectors from "./selectors";

import { initialState } from "./reducers/initialState";
import BaseService from "./services/BaseService";
export * from "./settings/core";
import { productToBasketItem as productToBasketItemTypes } from "./models/feature/feature.types";
// uses "configureStore" from ./store, should be imported after ./store
import * as ReduxTestUtils from "./utils/ReduxTestUtils";

export {
	decorators,
	Selectors,
	actions,
	reducers,
	DevTools,
	StoreConfig,
	InterceptReducers,
	AppStore,
	initialState,
	AppActions,
	AppState,
	connectFluxRedux,
	configureStore,
	getDefaultFluxStoreMap,
	ReduxTestUtils,
	channelUtils,
	actionTypes,
	BaseService,
	productToBasketItemTypes,
};

export * from "./actions/actionsPayload";
export * from "./services/Rest";
export * from "./services";
export * from "./types";
export * from "./decorators";
export * from "./models/msisdnSelection/msisdnSelection.actions";
export * from "./models/productOfferingConfiguration/productOfferingConfiguration.utils";
