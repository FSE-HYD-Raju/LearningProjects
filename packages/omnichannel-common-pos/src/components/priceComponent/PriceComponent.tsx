import * as React from "react";
import { isNumber } from "lodash";
import { PriceType, SimplePrice } from "../../redux/types";
import ProductOfferingUtil from "../../utils/ProductOfferingUtil";
import { PriceTypeEnum, PriceRange } from "../../redux";
import { PriceComponentPriceType, PriceInfo } from "./model/PriceComponentPriceType";
import { Configurations, ProductOffering } from "../../redux/types";
import { withProductConfigurations, ProductConfigurationInputProps } from "../product/withProductConfigurations";
import messages, { PriceComponentMessagesType } from "./PriceComponent.messages";
import { FormattedMessageDescriptor } from "../../channelUtils";

interface PriceComponentProps extends ProductConfigurationInputProps {
	configurations: Configurations;
	renderer: React.ComponentType<PriceRendererProps>;
	product: ProductOffering;
	// defines a function that should be used for calculating price of the offering
	priceExtractor?: (product: ProductOffering, priceType: PriceType, priceAttribute: string) => PriceInfo;
	priceAttribute: string;
}

interface PriceRendererProps {
	configurations: Configurations;
	prices: PriceComponentPriceType[];
	priceValueCalculator: (price: PriceComponentPriceType) => number;
	product: ProductOffering;
	messages: PriceComponentMessagesType;
}

// defines the way the price is transformed before passing to <FormattedNumber> inside renderer
const getPriceValue = (price: PriceComponentPriceType): number => {
	const negativityMultiplicator = +price.integer >= 0 ? 1 : -1;
	return parseInt(price.integer, 10) + negativityMultiplicator * parseInt(price.fraction, 10) * 0.01;
};

	// in case we are dealing with price range - use its "min" value, otherwise use taxFreeAmount

const pickPrice = (price: PriceInfo, priceAttribute: string): number | undefined => {
	const pick = price ? (isNumber((price as PriceRange).min)
		? (price as PriceRange).min
		: (price as SimplePrice)[priceAttribute]) : undefined;
	return pick;
};

/**
 * A component capable of rendering prices with appropriate i18ized messages based on renderer provided to B2CCheckoutDeliveryIdentificationProps
 *
 * Renderer should be a React.Component<PriceRendererProps>
 *
 */
const PriceComponent: React.FC<PriceComponentProps> = (props: PriceComponentProps) => {
	if (!props.renderer) {
		window.console.warn("No Renderer provided to PriceComponent");
		return null;
	}

	// the "extractor" function should be provided via props, defaults to ProductOfferingUtil.getPrice(...)
	// to get price range instead one should use ProductOfferingUtil.getPriceRange(..) but that ignores additional configuration
	const priceExtractor = props.priceExtractor || ProductOfferingUtil.getPrice;

	const { priceAttribute } = props;

	const upFrontPrice: PriceInfo = priceExtractor(props.product, PriceTypeEnum.ONE_TIME, priceAttribute);
	const recurringPrice: PriceInfo = priceExtractor(props.product, PriceTypeEnum.RECURRENT, priceAttribute);

	const upFrontSource = pickPrice(upFrontPrice, priceAttribute);

	const recurringSource = pickPrice(recurringPrice, priceAttribute);

	const prices: PriceComponentPriceType[] = [];

	// TODO: this should probably re-implemented, since we have number formatting in renderers too
	if (upFrontSource) {
		const formattedUpfront = {
			integer: upFrontSource.toFixed(2).split(".")[0],
			fraction: upFrontSource.toFixed(2).split(".")[1],
			currency: upFrontPrice.currency || "",
			priceType: "upfront",
			simplePrice: undefined,
			originalPrice: undefined
		};
		// FIXME: check if there is a need to display price if it is equal to zero or undefined
		prices.push(formattedUpfront);
	}

	if (recurringSource) {
		const formattedRecurring = {
			integer: recurringSource.toFixed(2).split(".")[0],
			fraction: recurringSource.toFixed(2).split(".")[1],
			currency: recurringPrice.currency || "",
			priceType: "recurring",
			simplePrice: undefined,
			originalPrice: undefined
		};
		prices.push(formattedRecurring);
	}

	const Renderer = props.renderer;
	return (
		<Renderer
			configurations={props.configurations}
			prices={prices}
			priceValueCalculator={getPriceValue}
			product={props.product}
			messages={messages}
		/>
	);
};

export default PriceComponent;
const PriceComponentWithProductConfiguration = withProductConfigurations<PriceComponentProps>(PriceComponent);
export { PriceComponentProps, PriceRendererProps, PriceInfo, PriceComponentWithProductConfiguration };
