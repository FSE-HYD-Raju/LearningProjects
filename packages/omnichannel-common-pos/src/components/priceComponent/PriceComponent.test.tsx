import * as React from "react";
import { mountWithContext, shallowWithContext } from "../../testUtils";
import ProductOfferingUtil from "../../utils/ProductOfferingUtil";
import { ProductOffering } from "../../redux/types";
import PriceComponent, { PriceComponentProps } from "./PriceComponent";
import GridPriceRenderer from "./renderers/GridPriceRenderer";
import StandardPriceRenderer from "./renderers/StandardPriceRenderer";

describe("PriceComponent", () => {
	const product: ProductOffering = {
		id: "myProduct",
		name: "myProductName",
		inputCharacteristics: {},
		selected: false,
	} as ProductOffering;

	const minProps: PriceComponentProps = {
		configurations: {},
		product: product,
		priceAttribute: "taxIncludedAmount",
		renderer: StandardPriceRenderer,
	};

	it("succeeds at shallow mount with minimal props", () => {
		const wrapper = shallowWithContext(<PriceComponent {...minProps}/>);
		expect(wrapper).toMatchSnapshot();
	});

	it("succeeds at deep mount with minimal props", () => {
		mountWithContext(<PriceComponent {...minProps}/>);
	});

	it("uses provided renderer", () => {
		const wrapper = mountWithContext(<PriceComponent {...minProps}/>);
		expect(wrapper.find("StandardPriceRenderer").length).toBeGreaterThan(0);
	});

	it("renders itself using proper renderer", () => {
		const props = {...minProps, renderer: GridPriceRenderer};
		const wrapper = mountWithContext(<PriceComponent {...props}/>);
		expect(wrapper.find("GridPriceRenderer").length).toEqual(1);
		expect(wrapper.find("StandardPriceRenderer").length).toEqual(0);
	});

	it("calls priceExtractor provided in props", () => {
		const extractor = jest.fn();
		const props = {...minProps, priceExtractor: extractor};
		mountWithContext(<PriceComponent {...props}/>);
		expect(extractor).toHaveBeenCalledTimes(2);
	});

	it("calls priceExtractor provided in props", () => {
		ProductOfferingUtil.getPrice = jest.fn();
		const props = {...minProps, priceExtractor: ProductOfferingUtil.getPrice};
		mountWithContext(<PriceComponent {...props}/>);
		expect(ProductOfferingUtil.getPrice).toHaveBeenCalledTimes(2);
	});

	describe("getPriceValue", () => {
		it("gives correct price from negative price", () => {
			const props = { ...minProps, renderer: GridPriceRenderer };
			const price = {
				currency: "EUR",
				fraction: "99",
				integer: "-19",
				originalPrice: undefined,
				priceType: "upfront",
				simplePrice: undefined,
			};
			const component = mountWithContext(<PriceComponent {...props}/>);
			const { priceValueCalculator } = component.find(GridPriceRenderer).props();
			expect(priceValueCalculator(price)).toEqual(-19.99);
		});

		it("gives correct price from positive price", () => {
			const props = { ...minProps, renderer: GridPriceRenderer };
			const price = {
				currency: "EUR",
				fraction: "99",
				integer: "19",
				originalPrice: undefined,
				priceType: "upfront",
				simplePrice: undefined,
			};
			const component = mountWithContext(<PriceComponent {...props}/>);
			const { priceValueCalculator } = component.find(GridPriceRenderer).props();
			expect(priceValueCalculator(price)).toEqual(19.99);
		});
	});
});
