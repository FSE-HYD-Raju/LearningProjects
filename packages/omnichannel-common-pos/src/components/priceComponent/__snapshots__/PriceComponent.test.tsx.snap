// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PriceComponent succeeds at shallow mount with minimal props 1`] = `
<StandardPriceRenderer
  configurations={Object {}}
  messages={
    Object {
      "mnth": Object {
        "defaultMessage": "/mth",
        "description": "Shop product price, per month price shorthand",
        "id": "shop-product-price-tag-mnth",
      },
      "monthly": Object {
        "defaultMessage": "monthly",
        "description": "Shop product price, per month price",
        "id": "shop-product-price-tag-monthly",
      },
      "now": Object {
        "defaultMessage": "now",
        "description": "Shop product price, price now",
        "id": "shop-product-price-tag-now",
      },
      "payNow": Object {
        "defaultMessage": "Pay now",
        "description": "Shop product price, price now (for single product page view)",
        "id": "shop-product-price-single-product-page-tag-now",
      },
      "recurring": Object {
        "defaultMessage": "Recurring fee",
        "description": "Shop product price, price recurring (for single product page view)",
        "id": "shop-product-price-single-product-page-tag-recurring",
      },
      "starting": Object {
        "defaultMessage": "Starting from",
        "description": "Shop product price, starting from label",
        "id": "shop-product-price-starting-from-label",
      },
    }
  }
  priceValueCalculator={[Function]}
  prices={Array []}
  product={
    Object {
      "id": "myProduct",
      "inputCharacteristics": Object {},
      "name": "myProductName",
      "selected": false,
    }
  }
/>
`;
